// @shubhu/pdfdiff/src/utils/visual-diff.js
/**
 * Visual diff utilities for PDF comparison with position data
 */

/**
 * @typedef {import('./pdfjs-extractor.js').PositionedTextItem} PositionedTextItem
 * @typedef {import('./pdfjs-extractor.js').PageTextContent} PageTextContent
 * @typedef {import('../types.js').VisualDiffItem} VisualDiffItem
 * @typedef {import('../types.js').VisualDiffResult} VisualDiffResult
 * @typedef {import('../types.js').VisualDiffType} VisualDiffType
 */

/**
 * Calculate distance between two text items based on position and content
 * @param {PositionedTextItem} item1 - First text item
 * @param {PositionedTextItem} item2 - Second text item
 * @returns {number} Distance score (lower is better match)
 */
function calculateItemDistance(item1, item2) {
  // Text content similarity (Levenshtein distance normalized)
  const textDistance = levenshteinDistance(item1.text, item2.text) / Math.max(item1.text.length, item2.text.length, 1);
  
  // Position distance (Euclidean distance normalized by page dimensions)
  const positionDistance = Math.sqrt(
    Math.pow(item1.x - item2.x, 2) + Math.pow(item1.y - item2.y, 2)
  ) / 1000; // Normalize by typical page size
  
  // Size similarity
  const sizeDistance = Math.abs(item1.width - item2.width) / Math.max(item1.width, item2.width, 1);
  
  // Combine distances with weights
  return textDistance * 0.6 + positionDistance * 0.3 + sizeDistance * 0.1;
}

/**
 * Simple Levenshtein distance implementation
 * @param {string} str1 - First string
 * @param {string} str2 - Second string
 * @returns {number} Edit distance
 */
function levenshteinDistance(str1, str2) {
  const matrix = [];
  
  for (let i = 0; i <= str2.length; i++) {
    matrix[i] = [i];
  }
  
  for (let j = 0; j <= str1.length; j++) {
    matrix[0][j] = j;
  }
  
  for (let i = 1; i <= str2.length; i++) {
    for (let j = 1; j <= str1.length; j++) {
      if (str2.charAt(i - 1) === str1.charAt(j - 1)) {
        matrix[i][j] = matrix[i - 1][j - 1];
      } else {
        matrix[i][j] = Math.min(
          matrix[i - 1][j - 1] + 1,
          matrix[i][j - 1] + 1,
          matrix[i - 1][j] + 1
        );
      }
    }
  }
  
  return matrix[str2.length][str1.length];
}

/**
 * Find best matching items between two arrays using Hungarian algorithm approximation
 * @param {Array<PositionedTextItem>} items1 - First set of items
 * @param {Array<PositionedTextItem>} items2 - Second set of items
 * @param {number} [threshold=0.7] - Distance threshold for considering items as matched
 * @returns {Object} Matching result with matched, added, and removed items
 */
function findMatchingItems(items1, items2, threshold = 0.7) {
  const matched = [];
  const added = [];
  const removed = [];
  const used2 = new Set();
  
  // For each item in first set, find best match in second set
  for (const item1 of items1) {
    let bestMatch = null;
    let bestDistance = Infinity;
    let bestIndex = -1;
    
    for (let i = 0; i < items2.length; i++) {
      if (used2.has(i)) continue;
      
      const item2 = items2[i];
      const distance = calculateItemDistance(item1, item2);
      
      if (distance < bestDistance && distance < threshold) {
        bestDistance = distance;
        bestMatch = item2;
        bestIndex = i;
      }
    }
    
    if (bestMatch) {
      matched.push({
        item1,
        item2: bestMatch,
        distance: bestDistance
      });
      used2.add(bestIndex);
    } else {
      removed.push(item1);
    }
  }
  
  // Remaining items in second set are additions
  for (let i = 0; i < items2.length; i++) {
    if (!used2.has(i)) {
      added.push(items2[i]);
    }
  }
  
  return { matched, added, removed };
}

/**
 * Create visual diff items from matching results
 * @param {Object} matchingResult - Result from findMatchingItems
 * @returns {Array<VisualDiffItem>} Array of visual diff items
 */
function createVisualDiffItems(matchingResult) {
  /** @type {Array<VisualDiffItem>} */
  const diffItems = [];
  
  // Add unchanged items (matched items with low distance)
  for (const match of matchingResult.matched) {
    if (match.distance < 0.1) {
      // Very similar items are considered unchanged
      diffItems.push({
        text: match.item1.text,
        x: match.item1.x,
        y: match.item1.y,
        width: match.item1.width,
        height: match.item1.height,
        page: match.item1.page,
        type: /** @type {VisualDiffType} */ ('unchanged'),
        fontName: match.item1.fontName,
        transform: match.item1.transform
      });
    } else {
      // Items with changes - show both removed and added
      diffItems.push({
        text: match.item1.text,
        x: match.item1.x,
        y: match.item1.y,
        width: match.item1.width,
        height: match.item1.height,
        page: match.item1.page,
        type: /** @type {VisualDiffType} */ ('removed'),
        fontName: match.item1.fontName,
        transform: match.item1.transform
      });
      
      diffItems.push({
        text: match.item2.text,
        x: match.item2.x,
        y: match.item2.y,
        width: match.item2.width,
        height: match.item2.height,
        page: match.item2.page,
        type: /** @type {VisualDiffType} */ ('added'),
        fontName: match.item2.fontName,
        transform: match.item2.transform
      });
    }
  }
  
  // Add removed items
  for (const item of matchingResult.removed) {
    diffItems.push({
      text: item.text,
      x: item.x,
      y: item.y,
      width: item.width,
      height: item.height,
      page: item.page,
      type: 'removed',
      fontName: item.fontName,
      transform: item.transform
    });
  }
  
  // Add added items
  for (const item of matchingResult.added) {
    diffItems.push({
      text: item.text,
      x: item.x,
      y: item.y,
      width: item.width,
      height: item.height,
      page: item.page,
      type: /** @type {VisualDiffType} */ ('added'),
      fontName: item.fontName,
      transform: item.transform
    });
  }
  
  return diffItems;
}

/**
 * Compare two PDFs visually using positioned text data
 * @param {Array<PageTextContent>} pdf1Content - First PDF's positioned text content
 * @param {Array<PageTextContent>} pdf2Content - Second PDF's positioned text content
 * @returns {VisualDiffResult} Visual diff result with positioned changes
 */
export function compareVisualPdfs(pdf1Content, pdf2Content) {
  const allDiffItems = [];
  const maxPages = Math.max(pdf1Content.length, pdf2Content.length);
  
  let totalAdded = 0;
  let totalRemoved = 0;
  let totalUnchanged = 0;
  
  // Compare each page
  for (let pageNum = 1; pageNum <= maxPages; pageNum++) {
    const page1 = pdf1Content.find(p => p.page === pageNum);
    const page2 = pdf2Content.find(p => p.page === pageNum);
    
    const items1 = page1 ? page1.items : [];
    const items2 = page2 ? page2.items : [];
    
    const matchingResult = findMatchingItems(items1, items2);
    const pageDiffItems = createVisualDiffItems(matchingResult);
    
    // Count changes for this page
    const pageAdded = pageDiffItems.filter(item => item.type === 'added').length;
    const pageRemoved = pageDiffItems.filter(item => item.type === 'removed').length;
    const pageUnchanged = pageDiffItems.filter(item => item.type === 'unchanged').length;
    
    totalAdded += pageAdded;
    totalRemoved += pageRemoved;
    totalUnchanged += pageUnchanged;
    
    allDiffItems.push(...pageDiffItems);
  }
  
  const identical = totalAdded === 0 && totalRemoved === 0;
  
  return {
    items: allDiffItems,
    summary: {
      addedCount: totalAdded,
      removedCount: totalRemoved,
      unchangedCount: totalUnchanged
    },
    identical,
    pageInfo: {
      maxPages
    }
  };
}

/**
 * Format visual diff result for overlay rendering
 * @param {VisualDiffResult} visualDiffResult - Visual diff result
 * @param {Object} [options] - Formatting options
 * @param {boolean} [options.showUnchanged=false] - Whether to include unchanged items
 * @param {Array<number>} [options.pages] - Specific pages to include (default: all)
 * @returns {Object} Formatted overlay data by page
 */
export function formatVisualDiffForOverlay(visualDiffResult, options = {}) {
  const { showUnchanged = false, pages } = options;
  const overlayData = {};
  
  // Filter items based on options
  let itemsToRender = visualDiffResult.items;
  
  if (!showUnchanged) {
    itemsToRender = itemsToRender.filter(item => item.type !== 'unchanged');
  }
  
  if (pages && pages.length > 0) {
    itemsToRender = itemsToRender.filter(item => pages.includes(item.page));
  }
  
  // Group by page
  for (const item of itemsToRender) {
    if (!overlayData[item.page]) {
      overlayData[item.page] = [];
    }
    
    overlayData[item.page].push({
      x: item.x,
      y: item.y,
      width: item.width,
      height: item.height,
      text: item.text,
      type: item.type,
      fontName: item.fontName
    });
  }
  
  return overlayData;
}

export { calculateItemDistance, findMatchingItems };
