// @shubhu/pdfdiff/src/browser.js
// Browser-compatible version of pdfdiff
// This is the primary entry point for browser usage

import { compareTexts, formatDiff } from './core.js';
import { extractPdfText as extractPdfTextWithPdfjs, extractPositionedPdfText } from './utils/pdfjs-extractor.js';
import { compareVisualPdfs as compareVisualPdfsImpl, formatVisualDiffForOverlay } from './utils/visual-diff.js';
import { toUint8Array } from './utils/pdf.js';

/**
 * Extract text from PDF data using pdfjs-dist (browser version)
 * @param {ArrayBuffer|Uint8Array|Buffer} pdfData - PDF data
 * @param {import('./utils/pdfjs-extractor.js').PDFJSLib} [pdfjsLib] - Optional pdfjs-dist instance (defaults to global pdfjsLib)
 * @returns {Promise<string>} Extracted text content
 */
export async function extractPdfText(pdfData, pdfjsLib) {
  try {
    const uint8Array = toUint8Array(pdfData);
    return await extractPdfTextWithPdfjs(uint8Array, pdfjsLib);
  } catch (error) {
    throw new Error(`Failed to extract text from PDF: ${error.message}`);
  }
}

/**
 * Compare two PDFs (browser version)
 * @param {ArrayBuffer|Uint8Array|Buffer|string} pdf1 - First PDF data or text content
 * @param {ArrayBuffer|Uint8Array|Buffer|string} pdf2 - Second PDF data or text content
 * @param {import('./types.js').BrowserCompareOptions} [options] - Comparison options
 * @returns {Promise<import('./types.js').DiffResult>} Comparison result
 */
export async function comparePdfs(pdf1, pdf2, options = {}) {
  const { ignoreWhitespace = false, ignoreCase = false, mode = 'char', pdfjsLib } = options;
  
  let text1, text2;
  
  // If inputs are strings, use them directly
  if (typeof pdf1 === 'string' && typeof pdf2 === 'string') {
    text1 = pdf1;
    text2 = pdf2;
  } else {
    // Extract text from PDF data
    text1 = await extractPdfText(/** @type {ArrayBuffer|Uint8Array|Buffer} */ (pdf1), pdfjsLib);
    text2 = await extractPdfText(/** @type {ArrayBuffer|Uint8Array|Buffer} */ (pdf2), pdfjsLib);
  }
  
  return compareTexts(text1, text2, { ignoreWhitespace, ignoreCase, mode });
}

/**
 * Extract positioned text from PDF data (browser version)
 * @param {ArrayBuffer|Uint8Array|Buffer} pdfData - PDF data
 * @param {import('./utils/pdfjs-extractor.js').PDFJSLib} [pdfjsLib] - Optional pdfjs-dist instance
 * @returns {Promise<Array<import('./utils/pdfjs-extractor.js').PageTextContent>>} Array of page text content with positions
 */
export async function extractPositionedText(pdfData, pdfjsLib) {
  try {
    const uint8Array = toUint8Array(pdfData);
    return await extractPositionedPdfText(uint8Array, pdfjsLib);
  } catch (error) {
    throw new Error(`Failed to extract positioned text from PDF: ${error.message}`);
  }
}

/**
 * Compare two PDFs visually with position information (browser version)
 * @param {ArrayBuffer|Uint8Array|Buffer} pdf1 - First PDF data
 * @param {ArrayBuffer|Uint8Array|Buffer} pdf2 - Second PDF data  
 * @param {Object} [options] - Comparison options
 * @param {import('./utils/pdfjs-extractor.js').PDFJSLib} [options.pdfjsLib] - pdfjs-dist instance for text extraction
 * @returns {Promise<import('./utils/visual-diff.js').VisualDiffResult>} Visual diff result with positioned changes
 */
export async function compareVisualPdfs(pdf1, pdf2, options = {}) {
  const { pdfjsLib } = options;
  
  try {
    // Extract positioned text from both PDFs
    const pdf1Content = await extractPositionedText(pdf1, pdfjsLib);
    const pdf2Content = await extractPositionedText(pdf2, pdfjsLib);
    
    // Perform visual comparison
    const visualDiff = compareVisualPdfsImpl(pdf1Content, pdf2Content);
    
    return visualDiff;
  } catch (error) {
    throw new Error(`Failed to compare PDFs visually: ${error.message}`);
  }
}

// Re-export utilities
export { compareTexts, formatDiff, formatVisualDiffForOverlay };

/**
 * @typedef {Object} PDFDiffGlobal
 * @property {function(ArrayBuffer|Uint8Array|Buffer, Object=): Promise<string>} extractPdfText - Extract text from PDF data
 * @property {function(ArrayBuffer|Uint8Array|Buffer, Object=): Promise<Array>} extractPositionedText - Extract positioned text from PDF data
 * @property {function(string, string, Object=): import('./types.js').DiffResult} compareTexts - Compare two text strings
 * @property {function(ArrayBuffer|Uint8Array|Buffer|string, ArrayBuffer|Uint8Array|Buffer|string, Object=): Promise<import('./types.js').DiffResult>} comparePdfs - Compare two PDF files
 * @property {function(ArrayBuffer|Uint8Array|Buffer, ArrayBuffer|Uint8Array|Buffer, Object=): Promise<Object>} compareVisualPdfs - Compare two PDFs visually with position data
 * @property {function(import('./types.js').DiffResult, Object=): Promise<string>} formatDiff - Format diff result for display
 * @property {function(Object, Object=): Object} formatVisualDiffForOverlay - Format visual diff for overlay rendering
 */

// Create a global object for browser usage
if (typeof window !== 'undefined') {
  /** @type {PDFDiffGlobal} */ 
  (/** @type {any} */ (window).PDFDiff) = {
    extractPdfText,
    extractPositionedText,
    compareTexts,
    comparePdfs,
    compareVisualPdfs,
    formatDiff,
    formatVisualDiffForOverlay
  };
}

export default { 
  extractPdfText,
  extractPositionedText, 
  compareTexts, 
  comparePdfs,
  compareVisualPdfs, 
  formatDiff,
  formatVisualDiffForOverlay 
};
