// @shubhu/pdfdiff/dist/pdfdiff.iife.js
var PDFDiff = (function (exports) {
    'use strict';

    class Diff {
        diff(oldStr, newStr, 
        // Type below is not accurate/complete - see above for full possibilities - but it compiles
        options = {}) {
            let callback;
            if (typeof options === 'function') {
                callback = options;
                options = {};
            }
            else if ('callback' in options) {
                callback = options.callback;
            }
            // Allow subclasses to massage the input prior to running
            const oldString = this.castInput(oldStr, options);
            const newString = this.castInput(newStr, options);
            const oldTokens = this.removeEmpty(this.tokenize(oldString, options));
            const newTokens = this.removeEmpty(this.tokenize(newString, options));
            return this.diffWithOptionsObj(oldTokens, newTokens, options, callback);
        }
        diffWithOptionsObj(oldTokens, newTokens, options, callback) {
            var _a;
            const done = (value) => {
                value = this.postProcess(value, options);
                if (callback) {
                    setTimeout(function () { callback(value); }, 0);
                    return undefined;
                }
                else {
                    return value;
                }
            };
            const newLen = newTokens.length, oldLen = oldTokens.length;
            let editLength = 1;
            let maxEditLength = newLen + oldLen;
            if (options.maxEditLength != null) {
                maxEditLength = Math.min(maxEditLength, options.maxEditLength);
            }
            const maxExecutionTime = (_a = options.timeout) !== null && _a !== void 0 ? _a : Infinity;
            const abortAfterTimestamp = Date.now() + maxExecutionTime;
            const bestPath = [{ oldPos: -1, lastComponent: undefined }];
            // Seed editLength = 0, i.e. the content starts with the same values
            let newPos = this.extractCommon(bestPath[0], newTokens, oldTokens, 0, options);
            if (bestPath[0].oldPos + 1 >= oldLen && newPos + 1 >= newLen) {
                // Identity per the equality and tokenizer
                return done(this.buildValues(bestPath[0].lastComponent, newTokens, oldTokens));
            }
            // Once we hit the right edge of the edit graph on some diagonal k, we can
            // definitely reach the end of the edit graph in no more than k edits, so
            // there's no point in considering any moves to diagonal k+1 any more (from
            // which we're guaranteed to need at least k+1 more edits).
            // Similarly, once we've reached the bottom of the edit graph, there's no
            // point considering moves to lower diagonals.
            // We record this fact by setting minDiagonalToConsider and
            // maxDiagonalToConsider to some finite value once we've hit the edge of
            // the edit graph.
            // This optimization is not faithful to the original algorithm presented in
            // Myers's paper, which instead pointlessly extends D-paths off the end of
            // the edit graph - see page 7 of Myers's paper which notes this point
            // explicitly and illustrates it with a diagram. This has major performance
            // implications for some common scenarios. For instance, to compute a diff
            // where the new text simply appends d characters on the end of the
            // original text of length n, the true Myers algorithm will take O(n+d^2)
            // time while this optimization needs only O(n+d) time.
            let minDiagonalToConsider = -Infinity, maxDiagonalToConsider = Infinity;
            // Main worker method. checks all permutations of a given edit length for acceptance.
            const execEditLength = () => {
                for (let diagonalPath = Math.max(minDiagonalToConsider, -editLength); diagonalPath <= Math.min(maxDiagonalToConsider, editLength); diagonalPath += 2) {
                    let basePath;
                    const removePath = bestPath[diagonalPath - 1], addPath = bestPath[diagonalPath + 1];
                    if (removePath) {
                        // No one else is going to attempt to use this value, clear it
                        // @ts-expect-error - perf optimisation. This type-violating value will never be read.
                        bestPath[diagonalPath - 1] = undefined;
                    }
                    let canAdd = false;
                    if (addPath) {
                        // what newPos will be after we do an insertion:
                        const addPathNewPos = addPath.oldPos - diagonalPath;
                        canAdd = addPath && 0 <= addPathNewPos && addPathNewPos < newLen;
                    }
                    const canRemove = removePath && removePath.oldPos + 1 < oldLen;
                    if (!canAdd && !canRemove) {
                        // If this path is a terminal then prune
                        // @ts-expect-error - perf optimisation. This type-violating value will never be read.
                        bestPath[diagonalPath] = undefined;
                        continue;
                    }
                    // Select the diagonal that we want to branch from. We select the prior
                    // path whose position in the old string is the farthest from the origin
                    // and does not pass the bounds of the diff graph
                    if (!canRemove || (canAdd && removePath.oldPos < addPath.oldPos)) {
                        basePath = this.addToPath(addPath, true, false, 0, options);
                    }
                    else {
                        basePath = this.addToPath(removePath, false, true, 1, options);
                    }
                    newPos = this.extractCommon(basePath, newTokens, oldTokens, diagonalPath, options);
                    if (basePath.oldPos + 1 >= oldLen && newPos + 1 >= newLen) {
                        // If we have hit the end of both strings, then we are done
                        return done(this.buildValues(basePath.lastComponent, newTokens, oldTokens)) || true;
                    }
                    else {
                        bestPath[diagonalPath] = basePath;
                        if (basePath.oldPos + 1 >= oldLen) {
                            maxDiagonalToConsider = Math.min(maxDiagonalToConsider, diagonalPath - 1);
                        }
                        if (newPos + 1 >= newLen) {
                            minDiagonalToConsider = Math.max(minDiagonalToConsider, diagonalPath + 1);
                        }
                    }
                }
                editLength++;
            };
            // Performs the length of edit iteration. Is a bit fugly as this has to support the
            // sync and async mode which is never fun. Loops over execEditLength until a value
            // is produced, or until the edit length exceeds options.maxEditLength (if given),
            // in which case it will return undefined.
            if (callback) {
                (function exec() {
                    setTimeout(function () {
                        if (editLength > maxEditLength || Date.now() > abortAfterTimestamp) {
                            return callback(undefined);
                        }
                        if (!execEditLength()) {
                            exec();
                        }
                    }, 0);
                }());
            }
            else {
                while (editLength <= maxEditLength && Date.now() <= abortAfterTimestamp) {
                    const ret = execEditLength();
                    if (ret) {
                        return ret;
                    }
                }
            }
        }
        addToPath(path, added, removed, oldPosInc, options) {
            const last = path.lastComponent;
            if (last && !options.oneChangePerToken && last.added === added && last.removed === removed) {
                return {
                    oldPos: path.oldPos + oldPosInc,
                    lastComponent: { count: last.count + 1, added: added, removed: removed, previousComponent: last.previousComponent }
                };
            }
            else {
                return {
                    oldPos: path.oldPos + oldPosInc,
                    lastComponent: { count: 1, added: added, removed: removed, previousComponent: last }
                };
            }
        }
        extractCommon(basePath, newTokens, oldTokens, diagonalPath, options) {
            const newLen = newTokens.length, oldLen = oldTokens.length;
            let oldPos = basePath.oldPos, newPos = oldPos - diagonalPath, commonCount = 0;
            while (newPos + 1 < newLen && oldPos + 1 < oldLen && this.equals(oldTokens[oldPos + 1], newTokens[newPos + 1], options)) {
                newPos++;
                oldPos++;
                commonCount++;
                if (options.oneChangePerToken) {
                    basePath.lastComponent = { count: 1, previousComponent: basePath.lastComponent, added: false, removed: false };
                }
            }
            if (commonCount && !options.oneChangePerToken) {
                basePath.lastComponent = { count: commonCount, previousComponent: basePath.lastComponent, added: false, removed: false };
            }
            basePath.oldPos = oldPos;
            return newPos;
        }
        equals(left, right, options) {
            if (options.comparator) {
                return options.comparator(left, right);
            }
            else {
                return left === right
                    || (!!options.ignoreCase && left.toLowerCase() === right.toLowerCase());
            }
        }
        removeEmpty(array) {
            const ret = [];
            for (let i = 0; i < array.length; i++) {
                if (array[i]) {
                    ret.push(array[i]);
                }
            }
            return ret;
        }
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        castInput(value, options) {
            return value;
        }
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        tokenize(value, options) {
            return Array.from(value);
        }
        join(chars) {
            // Assumes ValueT is string, which is the case for most subclasses.
            // When it's false, e.g. in diffArrays, this method needs to be overridden (e.g. with a no-op)
            // Yes, the casts are verbose and ugly, because this pattern - of having the base class SORT OF
            // assume tokens and values are strings, but not completely - is weird and janky.
            return chars.join('');
        }
        postProcess(changeObjects, 
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        options) {
            return changeObjects;
        }
        get useLongestToken() {
            return false;
        }
        buildValues(lastComponent, newTokens, oldTokens) {
            // First we convert our linked list of components in reverse order to an
            // array in the right order:
            const components = [];
            let nextComponent;
            while (lastComponent) {
                components.push(lastComponent);
                nextComponent = lastComponent.previousComponent;
                delete lastComponent.previousComponent;
                lastComponent = nextComponent;
            }
            components.reverse();
            const componentLen = components.length;
            let componentPos = 0, newPos = 0, oldPos = 0;
            for (; componentPos < componentLen; componentPos++) {
                const component = components[componentPos];
                if (!component.removed) {
                    if (!component.added && this.useLongestToken) {
                        let value = newTokens.slice(newPos, newPos + component.count);
                        value = value.map(function (value, i) {
                            const oldValue = oldTokens[oldPos + i];
                            return oldValue.length > value.length ? oldValue : value;
                        });
                        component.value = this.join(value);
                    }
                    else {
                        component.value = this.join(newTokens.slice(newPos, newPos + component.count));
                    }
                    newPos += component.count;
                    // Common case
                    if (!component.added) {
                        oldPos += component.count;
                    }
                }
                else {
                    component.value = this.join(oldTokens.slice(oldPos, oldPos + component.count));
                    oldPos += component.count;
                }
            }
            return components;
        }
    }

    class CharacterDiff extends Diff {
    }
    const characterDiff = new CharacterDiff();
    function diffChars(oldStr, newStr, options) {
        return characterDiff.diff(oldStr, newStr, options);
    }

    function longestCommonPrefix(str1, str2) {
        let i;
        for (i = 0; i < str1.length && i < str2.length; i++) {
            if (str1[i] != str2[i]) {
                return str1.slice(0, i);
            }
        }
        return str1.slice(0, i);
    }
    function longestCommonSuffix(str1, str2) {
        let i;
        // Unlike longestCommonPrefix, we need a special case to handle all scenarios
        // where we return the empty string since str1.slice(-0) will return the
        // entire string.
        if (!str1 || !str2 || str1[str1.length - 1] != str2[str2.length - 1]) {
            return '';
        }
        for (i = 0; i < str1.length && i < str2.length; i++) {
            if (str1[str1.length - (i + 1)] != str2[str2.length - (i + 1)]) {
                return str1.slice(-i);
            }
        }
        return str1.slice(-i);
    }
    function replacePrefix(string, oldPrefix, newPrefix) {
        if (string.slice(0, oldPrefix.length) != oldPrefix) {
            throw Error(`string ${JSON.stringify(string)} doesn't start with prefix ${JSON.stringify(oldPrefix)}; this is a bug`);
        }
        return newPrefix + string.slice(oldPrefix.length);
    }
    function replaceSuffix(string, oldSuffix, newSuffix) {
        if (!oldSuffix) {
            return string + newSuffix;
        }
        if (string.slice(-oldSuffix.length) != oldSuffix) {
            throw Error(`string ${JSON.stringify(string)} doesn't end with suffix ${JSON.stringify(oldSuffix)}; this is a bug`);
        }
        return string.slice(0, -oldSuffix.length) + newSuffix;
    }
    function removePrefix(string, oldPrefix) {
        return replacePrefix(string, oldPrefix, '');
    }
    function removeSuffix(string, oldSuffix) {
        return replaceSuffix(string, oldSuffix, '');
    }
    function maximumOverlap(string1, string2) {
        return string2.slice(0, overlapCount(string1, string2));
    }
    // Nicked from https://stackoverflow.com/a/60422853/1709587
    function overlapCount(a, b) {
        // Deal with cases where the strings differ in length
        let startA = 0;
        if (a.length > b.length) {
            startA = a.length - b.length;
        }
        let endB = b.length;
        if (a.length < b.length) {
            endB = a.length;
        }
        // Create a back-reference for each index
        //   that should be followed in case of a mismatch.
        //   We only need B to make these references:
        const map = Array(endB);
        let k = 0; // Index that lags behind j
        map[0] = 0;
        for (let j = 1; j < endB; j++) {
            if (b[j] == b[k]) {
                map[j] = map[k]; // skip over the same character (optional optimisation)
            }
            else {
                map[j] = k;
            }
            while (k > 0 && b[j] != b[k]) {
                k = map[k];
            }
            if (b[j] == b[k]) {
                k++;
            }
        }
        // Phase 2: use these references while iterating over A
        k = 0;
        for (let i = startA; i < a.length; i++) {
            while (k > 0 && a[i] != b[k]) {
                k = map[k];
            }
            if (a[i] == b[k]) {
                k++;
            }
        }
        return k;
    }
    function trailingWs(string) {
        // Yes, this looks overcomplicated and dumb - why not replace the whole function with
        //     return string match(/\s*$/)[0]
        // you ask? Because:
        // 1. the trap described at https://markamery.com/blog/quadratic-time-regexes/ would mean doing
        //    this would cause this function to take O(n²) time in the worst case (specifically when
        //    there is a massive run of NON-TRAILING whitespace in `string`), and
        // 2. the fix proposed in the same blog post, of using a negative lookbehind, is incompatible
        //    with old Safari versions that we'd like to not break if possible (see
        //    https://github.com/kpdecker/jsdiff/pull/550)
        // It feels absurd to do this with an explicit loop instead of a regex, but I really can't see a
        // better way that doesn't result in broken behaviour.
        let i;
        for (i = string.length - 1; i >= 0; i--) {
            if (!string[i].match(/\s/)) {
                break;
            }
        }
        return string.substring(i + 1);
    }
    function leadingWs(string) {
        // Thankfully the annoying considerations described in trailingWs don't apply here:
        const match = string.match(/^\s*/);
        return match ? match[0] : '';
    }

    // Based on https://en.wikipedia.org/wiki/Latin_script_in_Unicode
    //
    // Ranges and exceptions:
    // Latin-1 Supplement, 0080–00FF
    //  - U+00D7  × Multiplication sign
    //  - U+00F7  ÷ Division sign
    // Latin Extended-A, 0100–017F
    // Latin Extended-B, 0180–024F
    // IPA Extensions, 0250–02AF
    // Spacing Modifier Letters, 02B0–02FF
    //  - U+02C7  ˇ &#711;  Caron
    //  - U+02D8  ˘ &#728;  Breve
    //  - U+02D9  ˙ &#729;  Dot Above
    //  - U+02DA  ˚ &#730;  Ring Above
    //  - U+02DB  ˛ &#731;  Ogonek
    //  - U+02DC  ˜ &#732;  Small Tilde
    //  - U+02DD  ˝ &#733;  Double Acute Accent
    // Latin Extended Additional, 1E00–1EFF
    const extendedWordChars = 'a-zA-Z0-9_\\u{C0}-\\u{FF}\\u{D8}-\\u{F6}\\u{F8}-\\u{2C6}\\u{2C8}-\\u{2D7}\\u{2DE}-\\u{2FF}\\u{1E00}-\\u{1EFF}';
    // Each token is one of the following:
    // - A punctuation mark plus the surrounding whitespace
    // - A word plus the surrounding whitespace
    // - Pure whitespace (but only in the special case where this the entire text
    //   is just whitespace)
    //
    // We have to include surrounding whitespace in the tokens because the two
    // alternative approaches produce horribly broken results:
    // * If we just discard the whitespace, we can't fully reproduce the original
    //   text from the sequence of tokens and any attempt to render the diff will
    //   get the whitespace wrong.
    // * If we have separate tokens for whitespace, then in a typical text every
    //   second token will be a single space character. But this often results in
    //   the optimal diff between two texts being a perverse one that preserves
    //   the spaces between words but deletes and reinserts actual common words.
    //   See https://github.com/kpdecker/jsdiff/issues/160#issuecomment-1866099640
    //   for an example.
    //
    // Keeping the surrounding whitespace of course has implications for .equals
    // and .join, not just .tokenize.
    // This regex does NOT fully implement the tokenization rules described above.
    // Instead, it gives runs of whitespace their own "token". The tokenize method
    // then handles stitching whitespace tokens onto adjacent word or punctuation
    // tokens.
    const tokenizeIncludingWhitespace = new RegExp(`[${extendedWordChars}]+|\\s+|[^${extendedWordChars}]`, 'ug');
    class WordDiff extends Diff {
        equals(left, right, options) {
            if (options.ignoreCase) {
                left = left.toLowerCase();
                right = right.toLowerCase();
            }
            return left.trim() === right.trim();
        }
        tokenize(value, options = {}) {
            let parts;
            if (options.intlSegmenter) {
                const segmenter = options.intlSegmenter;
                if (segmenter.resolvedOptions().granularity != 'word') {
                    throw new Error('The segmenter passed must have a granularity of "word"');
                }
                parts = Array.from(segmenter.segment(value), segment => segment.segment);
            }
            else {
                parts = value.match(tokenizeIncludingWhitespace) || [];
            }
            const tokens = [];
            let prevPart = null;
            parts.forEach(part => {
                if ((/\s/).test(part)) {
                    if (prevPart == null) {
                        tokens.push(part);
                    }
                    else {
                        tokens.push(tokens.pop() + part);
                    }
                }
                else if (prevPart != null && (/\s/).test(prevPart)) {
                    if (tokens[tokens.length - 1] == prevPart) {
                        tokens.push(tokens.pop() + part);
                    }
                    else {
                        tokens.push(prevPart + part);
                    }
                }
                else {
                    tokens.push(part);
                }
                prevPart = part;
            });
            return tokens;
        }
        join(tokens) {
            // Tokens being joined here will always have appeared consecutively in the
            // same text, so we can simply strip off the leading whitespace from all the
            // tokens except the first (and except any whitespace-only tokens - but such
            // a token will always be the first and only token anyway) and then join them
            // and the whitespace around words and punctuation will end up correct.
            return tokens.map((token, i) => {
                if (i == 0) {
                    return token;
                }
                else {
                    return token.replace((/^\s+/), '');
                }
            }).join('');
        }
        postProcess(changes, options) {
            if (!changes || options.oneChangePerToken) {
                return changes;
            }
            let lastKeep = null;
            // Change objects representing any insertion or deletion since the last
            // "keep" change object. There can be at most one of each.
            let insertion = null;
            let deletion = null;
            changes.forEach(change => {
                if (change.added) {
                    insertion = change;
                }
                else if (change.removed) {
                    deletion = change;
                }
                else {
                    if (insertion || deletion) { // May be false at start of text
                        dedupeWhitespaceInChangeObjects(lastKeep, deletion, insertion, change);
                    }
                    lastKeep = change;
                    insertion = null;
                    deletion = null;
                }
            });
            if (insertion || deletion) {
                dedupeWhitespaceInChangeObjects(lastKeep, deletion, insertion, null);
            }
            return changes;
        }
    }
    const wordDiff = new WordDiff();
    function diffWords(oldStr, newStr, options) {
        return wordDiff.diff(oldStr, newStr, options);
    }
    function dedupeWhitespaceInChangeObjects(startKeep, deletion, insertion, endKeep) {
        // Before returning, we tidy up the leading and trailing whitespace of the
        // change objects to eliminate cases where trailing whitespace in one object
        // is repeated as leading whitespace in the next.
        // Below are examples of the outcomes we want here to explain the code.
        // I=insert, K=keep, D=delete
        // 1. diffing 'foo bar baz' vs 'foo baz'
        //    Prior to cleanup, we have K:'foo ' D:' bar ' K:' baz'
        //    After cleanup, we want:   K:'foo ' D:'bar ' K:'baz'
        //
        // 2. Diffing 'foo bar baz' vs 'foo qux baz'
        //    Prior to cleanup, we have K:'foo ' D:' bar ' I:' qux ' K:' baz'
        //    After cleanup, we want K:'foo ' D:'bar' I:'qux' K:' baz'
        //
        // 3. Diffing 'foo\nbar baz' vs 'foo baz'
        //    Prior to cleanup, we have K:'foo ' D:'\nbar ' K:' baz'
        //    After cleanup, we want K'foo' D:'\nbar' K:' baz'
        //
        // 4. Diffing 'foo baz' vs 'foo\nbar baz'
        //    Prior to cleanup, we have K:'foo\n' I:'\nbar ' K:' baz'
        //    After cleanup, we ideally want K'foo' I:'\nbar' K:' baz'
        //    but don't actually manage this currently (the pre-cleanup change
        //    objects don't contain enough information to make it possible).
        //
        // 5. Diffing 'foo   bar baz' vs 'foo  baz'
        //    Prior to cleanup, we have K:'foo  ' D:'   bar ' K:'  baz'
        //    After cleanup, we want K:'foo  ' D:' bar ' K:'baz'
        //
        // Our handling is unavoidably imperfect in the case where there's a single
        // indel between keeps and the whitespace has changed. For instance, consider
        // diffing 'foo\tbar\nbaz' vs 'foo baz'. Unless we create an extra change
        // object to represent the insertion of the space character (which isn't even
        // a token), we have no way to avoid losing information about the texts'
        // original whitespace in the result we return. Still, we do our best to
        // output something that will look sensible if we e.g. print it with
        // insertions in green and deletions in red.
        // Between two "keep" change objects (or before the first or after the last
        // change object), we can have either:
        // * A "delete" followed by an "insert"
        // * Just an "insert"
        // * Just a "delete"
        // We handle the three cases separately.
        if (deletion && insertion) {
            const oldWsPrefix = leadingWs(deletion.value);
            const oldWsSuffix = trailingWs(deletion.value);
            const newWsPrefix = leadingWs(insertion.value);
            const newWsSuffix = trailingWs(insertion.value);
            if (startKeep) {
                const commonWsPrefix = longestCommonPrefix(oldWsPrefix, newWsPrefix);
                startKeep.value = replaceSuffix(startKeep.value, newWsPrefix, commonWsPrefix);
                deletion.value = removePrefix(deletion.value, commonWsPrefix);
                insertion.value = removePrefix(insertion.value, commonWsPrefix);
            }
            if (endKeep) {
                const commonWsSuffix = longestCommonSuffix(oldWsSuffix, newWsSuffix);
                endKeep.value = replacePrefix(endKeep.value, newWsSuffix, commonWsSuffix);
                deletion.value = removeSuffix(deletion.value, commonWsSuffix);
                insertion.value = removeSuffix(insertion.value, commonWsSuffix);
            }
        }
        else if (insertion) {
            // The whitespaces all reflect what was in the new text rather than
            // the old, so we essentially have no information about whitespace
            // insertion or deletion. We just want to dedupe the whitespace.
            // We do that by having each change object keep its trailing
            // whitespace and deleting duplicate leading whitespace where
            // present.
            if (startKeep) {
                const ws = leadingWs(insertion.value);
                insertion.value = insertion.value.substring(ws.length);
            }
            if (endKeep) {
                const ws = leadingWs(endKeep.value);
                endKeep.value = endKeep.value.substring(ws.length);
            }
            // otherwise we've got a deletion and no insertion
        }
        else if (startKeep && endKeep) {
            const newWsFull = leadingWs(endKeep.value), delWsStart = leadingWs(deletion.value), delWsEnd = trailingWs(deletion.value);
            // Any whitespace that comes straight after startKeep in both the old and
            // new texts, assign to startKeep and remove from the deletion.
            const newWsStart = longestCommonPrefix(newWsFull, delWsStart);
            deletion.value = removePrefix(deletion.value, newWsStart);
            // Any whitespace that comes straight before endKeep in both the old and
            // new texts, and hasn't already been assigned to startKeep, assign to
            // endKeep and remove from the deletion.
            const newWsEnd = longestCommonSuffix(removePrefix(newWsFull, newWsStart), delWsEnd);
            deletion.value = removeSuffix(deletion.value, newWsEnd);
            endKeep.value = replacePrefix(endKeep.value, newWsFull, newWsEnd);
            // If there's any whitespace from the new text that HASN'T already been
            // assigned, assign it to the start:
            startKeep.value = replaceSuffix(startKeep.value, newWsFull, newWsFull.slice(0, newWsFull.length - newWsEnd.length));
        }
        else if (endKeep) {
            // We are at the start of the text. Preserve all the whitespace on
            // endKeep, and just remove whitespace from the end of deletion to the
            // extent that it overlaps with the start of endKeep.
            const endKeepWsPrefix = leadingWs(endKeep.value);
            const deletionWsSuffix = trailingWs(deletion.value);
            const overlap = maximumOverlap(deletionWsSuffix, endKeepWsPrefix);
            deletion.value = removeSuffix(deletion.value, overlap);
        }
        else if (startKeep) {
            // We are at the END of the text. Preserve all the whitespace on
            // startKeep, and just remove whitespace from the start of deletion to
            // the extent that it overlaps with the end of startKeep.
            const startKeepWsSuffix = trailingWs(startKeep.value);
            const deletionWsPrefix = leadingWs(deletion.value);
            const overlap = maximumOverlap(startKeepWsSuffix, deletionWsPrefix);
            deletion.value = removePrefix(deletion.value, overlap);
        }
    }

    class LineDiff extends Diff {
        constructor() {
            super(...arguments);
            this.tokenize = tokenize;
        }
        equals(left, right, options) {
            // If we're ignoring whitespace, we need to normalise lines by stripping
            // whitespace before checking equality. (This has an annoying interaction
            // with newlineIsToken that requires special handling: if newlines get their
            // own token, then we DON'T want to trim the *newline* tokens down to empty
            // strings, since this would cause us to treat whitespace-only line content
            // as equal to a separator between lines, which would be weird and
            // inconsistent with the documented behavior of the options.)
            if (options.ignoreWhitespace) {
                if (!options.newlineIsToken || !left.includes('\n')) {
                    left = left.trim();
                }
                if (!options.newlineIsToken || !right.includes('\n')) {
                    right = right.trim();
                }
            }
            else if (options.ignoreNewlineAtEof && !options.newlineIsToken) {
                if (left.endsWith('\n')) {
                    left = left.slice(0, -1);
                }
                if (right.endsWith('\n')) {
                    right = right.slice(0, -1);
                }
            }
            return super.equals(left, right, options);
        }
    }
    const lineDiff = new LineDiff();
    function diffLines(oldStr, newStr, options) {
        return lineDiff.diff(oldStr, newStr, options);
    }
    // Exported standalone so it can be used from jsonDiff too.
    function tokenize(value, options) {
        if (options.stripTrailingCr) {
            // remove one \r before \n to match GNU diff's --strip-trailing-cr behavior
            value = value.replace(/\r\n/g, '\n');
        }
        const retLines = [], linesAndNewlines = value.split(/(\n|\r\n)/);
        // Ignore the final empty token that occurs if the string ends with a new line
        if (!linesAndNewlines[linesAndNewlines.length - 1]) {
            linesAndNewlines.pop();
        }
        // Merge the content and line separators into single tokens
        for (let i = 0; i < linesAndNewlines.length; i++) {
            const line = linesAndNewlines[i];
            if (i % 2 && !options.newlineIsToken) {
                retLines[retLines.length - 1] += line;
            }
            else {
                retLines.push(line);
            }
        }
        return retLines;
    }

    /**
     * Apply text processing options
     * @param {string} text - Text to process
     * @param {import('../types.js').TextProcessingOptions} options - Processing options
     * @returns {string} Processed text
     */
    function processText(text, options = {}) {
      const { ignoreCase = false, ignoreWhitespace = false } = options;
      
      let processed = text;
      
      if (ignoreCase) {
        processed = processed.toLowerCase();
      }
      
      if (ignoreWhitespace) {
        processed = processed.replace(/\s+/g, ' ').trim();
      }
      
      return processed;
    }

    /**
     * Compare two texts and return differences
     * @param {string} text1 - First text content
     * @param {string} text2 - Second text content  
     * @param {import('../types.js').CompareOptions} [options] - Comparison options
     * @returns {import('../types.js').DiffResult} Comparison result
     */
    function compareTexts(text1, text2, options = {}) {
      const { ignoreWhitespace = false, ignoreCase = false, mode = 'char' } = options;
      
      try {
        const processedText1 = processText(text1, { ignoreCase, ignoreWhitespace });
        const processedText2 = processText(text2, { ignoreCase, ignoreWhitespace });
        
        // Choose diff function based on mode
        let changes;
        switch (mode) {
          case 'char':
            changes = diffChars(processedText1, processedText2);
            break;
          case 'word':
            changes = diffWords(processedText1, processedText2);
            break;
          case 'line':
            changes = diffLines(processedText1, processedText2);
            break;
          default:
            throw new Error(`Invalid diff mode: ${mode}. Must be 'char', 'word', or 'line'.`);
        }
        
        // Calculate summary statistics
        let added = 0;
        let removed = 0;
        
        changes.forEach(change => {
          if (change.added) {
            added += change.count || 1;
          } else if (change.removed) {
            removed += change.count || 1;
          }
        });
        
        const identical = changes.length === 0 || 
          (changes.length === 1 && !changes[0].added && !changes[0].removed);
        
        // Create mode-appropriate summary
        const unit = mode === 'char' ? 'characters' : mode === 'word' ? 'words' : 'lines';
        const summary = identical 
          ? 'PDFs are identical'
          : `${added} ${unit} added, ${removed} ${unit} removed`;
        
        return {
          changes,
          summary,
          identical
        };
      } catch (error) {
        throw new Error(`Failed to compare texts: ${error.message}`);
      }
    }

    /**
     * Handle different possible return formats from PDF extraction
     * @param {any} result - Result from PDF extraction function
     * @returns {string} Normalized text content
     */

    /**
     * Convert various input types to Uint8Array for consistent processing
     * @param {ArrayBuffer|Uint8Array|Buffer} data - Input data
     * @returns {Uint8Array} Converted data
     */
    function toUint8Array(data) {
      if (data instanceof ArrayBuffer) {
        return new Uint8Array(data);
      } else if (data instanceof Uint8Array) {
        return data;
      } else if (typeof Buffer !== 'undefined' && Buffer.isBuffer(data)) {
        return new Uint8Array(data);
      } else {
        throw new Error('Invalid input: expected ArrayBuffer, Uint8Array, or Buffer');
      }
    }

    let chalk;

    async function loadChalk() {
      if (chalk) return chalk;
      
      try {
        chalk = (await import('chalk')).default;
      } catch {
        // Fallback when chalk is not available (e.g., in browser environments)
        chalk = {
          green: (/** @type {string} */ text) => text,
          red: (/** @type {string} */ text) => text,
          blue: (/** @type {string} */ text) => text,
          dim: (/** @type {string} */ text) => text
        };
      }
      return chalk;
    }

    /**
     * Format diff output for display
     * @param {import('../types.js').DiffResult} diffResult - Result from comparePdfs or compareTexts
     * @param {import('../types.js').FormatOptions} [options] - Formatting options
     * @returns {Promise<string>} Formatted diff output
     */
    async function formatDiff(diffResult, options = {}) {
      const { showUnchanged = false, context = 3, color = true } = options;
      const { changes, summary, identical } = diffResult;
      
      const chalk = await loadChalk();
      
      if (identical) {
        const checkMark = color ? chalk.green('✓') : '✓';
        return `${checkMark} ${summary}\n`;
      }
      
      const fileIcon = color ? chalk.blue('📄') : '📄';
      let output = `${fileIcon} ${summary}\n\n`;
      
      changes.forEach((change) => {
        if (change.added) {
          const lines = change.value.split('\n');
          lines.forEach(line => {
            if (line.trim()) {
              const prefix = color ? chalk.green('+ ') : '+ ';
              const content = color ? chalk.green(line) : line;
              output += `${prefix}${content}\n`;
            }
          });
        } else if (change.removed) {
          const lines = change.value.split('\n');
          lines.forEach(line => {
            if (line.trim()) {
              const prefix = color ? chalk.red('- ') : '- ';
              const content = color ? chalk.red(line) : line;
              output += `${prefix}${content}\n`;
            }
          });
        } else {
          // Unchanged content - show normally when showUnchanged is true
          if (showUnchanged) {
            const lines = change.value.split('\n');
            if (lines.length > context * 2) {
              // Show first few lines
              for (let i = 0; i < context; i++) {
                if (lines[i]) {
                  const content = color ? chalk.dim(`  ${lines[i]}`) : `  ${lines[i]}`;
                  output += `${content}\n`;
                }
              }
              const dots = color ? chalk.dim('  ...') : '  ...';
              output += `${dots}\n`;
              // Show last few lines
              for (let i = lines.length - context; i < lines.length; i++) {
                if (lines[i]) {
                  const content = color ? chalk.dim(`  ${lines[i]}`) : `  ${lines[i]}`;
                  output += `${content}\n`;
                }
              }
            } else {
              lines.forEach(line => {
                if (line) {
                  const content = color ? chalk.dim(`  ${line}`) : `  ${line}`;
                  output += `${content}\n`;
                }
              });
            }
          }
        }
      });
      
      return output;
    }

    /**
     * PDF text extraction using pdfjs-dist
     * This module provides both Node.js and browser-compatible text extraction
     */

    /**
     * @typedef {Object} PDFDocumentLoadingParams
     * @property {Uint8Array} data - PDF data
     * @property {boolean} [disableRange] - Disable range requests
     * @property {boolean} [disableStream] - Disable stream parsing
     * @property {boolean} [disableAutoFetch] - Disable auto fetch
     */

    /**
     * @typedef {Object} PDFJSLib
     * @property {function(PDFDocumentLoadingParams): LoadingTask} getDocument - Get PDF document from data
     * @property {Object} GlobalWorkerOptions - Worker configuration
     */

    /**
     * @typedef {Object} PDFDocument
     * @property {number} numPages - Number of pages in the PDF
     * @property {function(number): Promise<PDFPage>} getPage - Get a specific page
     */

    /**
     * @typedef {Object} PDFPage
     * @property {function(): Promise<TextContent>} getTextContent - Get text content from page
     * @property {function(Object): Object} getViewport - Get page viewport with scale
     */

    /**
     * @typedef {Object} TextContent
     * @property {Array<TextItem>} items - Array of text items
     */

    /**
     * @typedef {Object} TextItem
     * @property {string} str - Text content
     * @property {Array<number>} transform - Text transformation matrix [a, b, c, d, e, f]
     * @property {number} width - Text width
     * @property {number} height - Text height
     * @property {number} [fontName] - Font name
     * @property {boolean} [hasEOL] - Has end of line
     */

    /**
     * @typedef {Object} PositionedTextItem
     * @property {string} text - Text content
     * @property {number} x - X coordinate
     * @property {number} y - Y coordinate  
     * @property {number} width - Text width
     * @property {number} height - Text height
     * @property {Array<number>} transform - Full transformation matrix
     * @property {string} [fontName] - Font name
     * @property {number} page - Page number (1-based)
     */

    /**
     * @typedef {Object} PageTextContent
     * @property {number} page - Page number
     * @property {Array<PositionedTextItem>} items - Positioned text items
     * @property {Object} viewport - Page viewport information
     */

    /**
     * @typedef {Object} LoadingTask
     * @property {Promise<PDFDocument>} promise - Promise that resolves to PDF document
     */

    /**
     * Extract text from PDF using pdfjs-dist
     * @param {Uint8Array} pdfData - PDF data as Uint8Array
     * @param {PDFJSLib} [pdfjsLib] - Optional pdfjs-dist instance (for browser use)
     * @returns {Promise<string>} Extracted text content
     */
    async function extractTextWithPdfjs(pdfData, pdfjsLib) {
      let pdfjs = pdfjsLib;
      
      // Try to load pdfjs-dist if not provided
      if (!pdfjs) {
        if (typeof window !== 'undefined' && /** @type {any} */ (window).pdfjsLib) {
          // Browser environment with global pdfjsLib
          pdfjs = /** @type {PDFJSLib} */ (/** @type {any} */ (window).pdfjsLib);
        } else if (typeof window === 'undefined') {
          // Node.js environment - use legacy build
          try {
            const pdfjsModule = await import('pdfjs-dist/legacy/build/pdf.mjs');
            pdfjs = /** @type {PDFJSLib} */ (/** @type {any} */ (pdfjsModule));
          } catch (error) {
            throw new Error(`Failed to load pdfjs-dist: ${error.message}`);
          }
        } else {
          throw new Error('No PDF.js library available. Please ensure pdfjsLib is available globally in browser environments.');
        }
      }
      
      try {
        // Load the PDF document
        const loadingTask = pdfjs.getDocument({ 
          data: pdfData,
          // Disable network requests and file system access for security
          disableRange: true,
          disableStream: true,
          disableAutoFetch: true
        });
        
        const pdfDocument = await loadingTask.promise;
        const numPages = pdfDocument.numPages;
        const textPages = [];
        
        // Extract text from each page
        for (let pageNum = 1; pageNum <= numPages; pageNum++) {
          const page = await pdfDocument.getPage(pageNum);
          const textContent = await page.getTextContent();
          
          // Combine all text items from the page
          const pageText = textContent.items
            .map(item => item.str)
            .join(' ');
          
          textPages.push(pageText);
        }
        
        // Join all pages with double newlines
        return textPages.join('\n\n');
        
      } catch (error) {
        throw new Error(`Failed to extract text using pdfjs-dist: ${error.message}`);
      }
    }

    /**
     * Extract text from PDF data with fallback handling
     * @param {ArrayBuffer|Uint8Array|Buffer} pdfData - PDF data
     * @param {PDFJSLib} [pdfjsLib] - Optional pdfjs-dist instance
     * @returns {Promise<string>} Extracted text content
     */
    async function extractPdfText$1(pdfData, pdfjsLib) {
      try {
        // Check if this is a mock PDF buffer for testing
        const textDecoder = new TextDecoder();
        try {
          const possibleJson = textDecoder.decode(pdfData);
          const mockData = JSON.parse(possibleJson);
          if (mockData.mockPdfText !== undefined) {
            return mockData.mockPdfText;
          }
        } catch {
          // Not a mock buffer, continue with normal PDF processing
        }

        // Convert to Uint8Array for consistent processing
        let uint8Array;
        if (pdfData instanceof ArrayBuffer) {
          uint8Array = new Uint8Array(pdfData);
        } else if (pdfData instanceof Uint8Array) {
          uint8Array = pdfData;
        } else if (typeof Buffer !== 'undefined' && Buffer.isBuffer(pdfData)) {
          uint8Array = new Uint8Array(pdfData);
        } else {
          throw new Error('Invalid input: expected ArrayBuffer, Uint8Array, or Buffer');
        }
        
        return await extractTextWithPdfjs(uint8Array, pdfjsLib);
      } catch (error) {
        throw new Error(`Failed to extract text from PDF: ${error.message}`);
      }
    }

    /**
     * Extract text with positioning information from PDF using pdfjs-dist
     * @param {Uint8Array} pdfData - PDF data as Uint8Array
     * @param {PDFJSLib} [pdfjsLib] - Optional pdfjs-dist instance (for browser use)
     * @returns {Promise<Array<PageTextContent>>} Array of page text content with positions
     */
    async function extractPositionedTextWithPdfjs(pdfData, pdfjsLib) {
      let pdfjs = pdfjsLib;
      
      // Try to load pdfjs-dist if not provided
      if (!pdfjs) {
        if (typeof window !== 'undefined' && /** @type {any} */ (window).pdfjsLib) {
          // Browser environment with global pdfjsLib
          pdfjs = /** @type {PDFJSLib} */ (/** @type {any} */ (window).pdfjsLib);
        } else if (typeof window === 'undefined') {
          // Node.js environment - use legacy build
          try {
            const pdfjsModule = await import('pdfjs-dist/legacy/build/pdf.mjs');
            pdfjs = /** @type {PDFJSLib} */ (/** @type {any} */ (pdfjsModule));
          } catch (error) {
            throw new Error(`Failed to load pdfjs-dist: ${error.message}`);
          }
        } else {
          throw new Error('No PDF.js library available. Please ensure pdfjsLib is available globally in browser environments.');
        }
      }
      
      try {
        // Load the PDF document
        const loadingTask = pdfjs.getDocument({ 
          data: pdfData,
          disableRange: true,
          disableStream: true,
          disableAutoFetch: true
        });
        
        const pdfDocument = await loadingTask.promise;
        const numPages = pdfDocument.numPages;
        const pageContents = [];
        
        // Extract positioned text from each page
        for (let pageNum = 1; pageNum <= numPages; pageNum++) {
          const page = await pdfDocument.getPage(pageNum);
          const viewport = page.getViewport({ scale: 1.0 });
          const textContent = await page.getTextContent();
          
          // Convert text items to positioned items
          const positionedItems = textContent.items.map((item) => {
            // Extract position from transformation matrix
            // Matrix format: [scaleX, skewY, skewX, scaleY, translateX, translateY]
            const transform = item.transform;
            const x = transform[4];
            const y = viewport.height - transform[5]; // Convert to top-down coordinates
            
            return {
              text: item.str,
              x: x,
              y: y,
              width: item.width,
              height: item.height || Math.abs(transform[0]), // Use scaleX as height fallback
              transform: transform,
              fontName: typeof item.fontName === 'string' ? item.fontName : 'unknown',
              page: pageNum
            };
          });
          
          pageContents.push({
            page: pageNum,
            items: positionedItems,
            viewport: {
              width: viewport.width,
              height: viewport.height
            }
          });
        }
        
        return pageContents;
        
      } catch (error) {
        throw new Error(`Failed to extract positioned text using pdfjs-dist: ${error.message}`);
      }
    }

    /**
     * Extract positioned text from PDF data with fallback handling
     * @param {ArrayBuffer|Uint8Array|Buffer} pdfData - PDF data
     * @param {PDFJSLib} [pdfjsLib] - Optional pdfjs-dist instance
     * @returns {Promise<Array<PageTextContent>>} Array of page text content with positions
     */
    async function extractPositionedPdfText(pdfData, pdfjsLib) {
      try {
        // Check if this is a mock PDF buffer for testing
        const textDecoder = new TextDecoder();
        try {
          const possibleJson = textDecoder.decode(pdfData);
          const mockData = JSON.parse(possibleJson);
          if (mockData.mockPositionedText !== undefined) {
            return mockData.mockPositionedText;
          }
        } catch {
          // Not a mock buffer, continue with normal PDF processing
        }

        // Convert to Uint8Array for consistent processing
        let uint8Array;
        if (pdfData instanceof ArrayBuffer) {
          uint8Array = new Uint8Array(pdfData);
        } else if (pdfData instanceof Uint8Array) {
          uint8Array = pdfData;
        } else if (typeof Buffer !== 'undefined' && Buffer.isBuffer(pdfData)) {
          uint8Array = new Uint8Array(pdfData);
        } else {
          throw new Error('Invalid input: expected ArrayBuffer, Uint8Array, or Buffer');
        }
        
        return await extractPositionedTextWithPdfjs(uint8Array, pdfjsLib);
      } catch (error) {
        throw new Error(`Failed to extract positioned text from PDF: ${error.message}`);
      }
    }

    /**
     * Visual diff utilities for PDF comparison with position data
     */

    /**
     * @typedef {import('./pdfjs-extractor.js').PositionedTextItem} PositionedTextItem
     * @typedef {import('./pdfjs-extractor.js').PageTextContent} PageTextContent
     * @typedef {import('../types.js').VisualDiffItem} VisualDiffItem
     * @typedef {import('../types.js').VisualDiffResult} VisualDiffResult
     * @typedef {import('../types.js').VisualDiffType} VisualDiffType
     */

    /**
     * Calculate distance between two text items based on position and content
     * @param {PositionedTextItem} item1 - First text item
     * @param {PositionedTextItem} item2 - Second text item
     * @returns {number} Distance score (lower is better match)
     */
    function calculateItemDistance(item1, item2) {
      // Text content similarity (Levenshtein distance normalized)
      const textDistance = levenshteinDistance(item1.text, item2.text) / Math.max(item1.text.length, item2.text.length, 1);
      
      // Position distance (Euclidean distance normalized by page dimensions)
      const positionDistance = Math.sqrt(
        Math.pow(item1.x - item2.x, 2) + Math.pow(item1.y - item2.y, 2)
      ) / 1000; // Normalize by typical page size
      
      // Size similarity
      const sizeDistance = Math.abs(item1.width - item2.width) / Math.max(item1.width, item2.width, 1);
      
      // Combine distances with weights
      return textDistance * 0.6 + positionDistance * 0.3 + sizeDistance * 0.1;
    }

    /**
     * Simple Levenshtein distance implementation
     * @param {string} str1 - First string
     * @param {string} str2 - Second string
     * @returns {number} Edit distance
     */
    function levenshteinDistance(str1, str2) {
      const matrix = [];
      
      for (let i = 0; i <= str2.length; i++) {
        matrix[i] = [i];
      }
      
      for (let j = 0; j <= str1.length; j++) {
        matrix[0][j] = j;
      }
      
      for (let i = 1; i <= str2.length; i++) {
        for (let j = 1; j <= str1.length; j++) {
          if (str2.charAt(i - 1) === str1.charAt(j - 1)) {
            matrix[i][j] = matrix[i - 1][j - 1];
          } else {
            matrix[i][j] = Math.min(
              matrix[i - 1][j - 1] + 1,
              matrix[i][j - 1] + 1,
              matrix[i - 1][j] + 1
            );
          }
        }
      }
      
      return matrix[str2.length][str1.length];
    }

    /**
     * Find best matching items between two arrays using Hungarian algorithm approximation
     * @param {Array<PositionedTextItem>} items1 - First set of items
     * @param {Array<PositionedTextItem>} items2 - Second set of items
     * @param {number} [threshold=0.7] - Distance threshold for considering items as matched
     * @returns {Object} Matching result with matched, added, and removed items
     */
    function findMatchingItems(items1, items2, threshold = 0.7) {
      const matched = [];
      const added = [];
      const removed = [];
      const used2 = new Set();
      
      // For each item in first set, find best match in second set
      for (const item1 of items1) {
        let bestMatch = null;
        let bestDistance = Infinity;
        let bestIndex = -1;
        
        for (let i = 0; i < items2.length; i++) {
          if (used2.has(i)) continue;
          
          const item2 = items2[i];
          const distance = calculateItemDistance(item1, item2);
          
          if (distance < bestDistance && distance < threshold) {
            bestDistance = distance;
            bestMatch = item2;
            bestIndex = i;
          }
        }
        
        if (bestMatch) {
          matched.push({
            item1,
            item2: bestMatch,
            distance: bestDistance
          });
          used2.add(bestIndex);
        } else {
          removed.push(item1);
        }
      }
      
      // Remaining items in second set are additions
      for (let i = 0; i < items2.length; i++) {
        if (!used2.has(i)) {
          added.push(items2[i]);
        }
      }
      
      return { matched, added, removed };
    }

    /**
     * Create visual diff items from matching results
     * @param {Object} matchingResult - Result from findMatchingItems
     * @returns {Array<VisualDiffItem>} Array of visual diff items
     */
    function createVisualDiffItems(matchingResult) {
      /** @type {Array<VisualDiffItem>} */
      const diffItems = [];
      
      // Add unchanged items (matched items with low distance)
      for (const match of matchingResult.matched) {
        if (match.distance < 0.1) {
          // Very similar items are considered unchanged
          diffItems.push({
            text: match.item1.text,
            x: match.item1.x,
            y: match.item1.y,
            width: match.item1.width,
            height: match.item1.height,
            page: match.item1.page,
            type: /** @type {VisualDiffType} */ ('unchanged'),
            fontName: match.item1.fontName,
            transform: match.item1.transform
          });
        } else {
          // Items with changes - show both removed and added
          diffItems.push({
            text: match.item1.text,
            x: match.item1.x,
            y: match.item1.y,
            width: match.item1.width,
            height: match.item1.height,
            page: match.item1.page,
            type: /** @type {VisualDiffType} */ ('removed'),
            fontName: match.item1.fontName,
            transform: match.item1.transform
          });
          
          diffItems.push({
            text: match.item2.text,
            x: match.item2.x,
            y: match.item2.y,
            width: match.item2.width,
            height: match.item2.height,
            page: match.item2.page,
            type: /** @type {VisualDiffType} */ ('added'),
            fontName: match.item2.fontName,
            transform: match.item2.transform
          });
        }
      }
      
      // Add removed items
      for (const item of matchingResult.removed) {
        diffItems.push({
          text: item.text,
          x: item.x,
          y: item.y,
          width: item.width,
          height: item.height,
          page: item.page,
          type: 'removed',
          fontName: item.fontName,
          transform: item.transform
        });
      }
      
      // Add added items
      for (const item of matchingResult.added) {
        diffItems.push({
          text: item.text,
          x: item.x,
          y: item.y,
          width: item.width,
          height: item.height,
          page: item.page,
          type: /** @type {VisualDiffType} */ ('added'),
          fontName: item.fontName,
          transform: item.transform
        });
      }
      
      return diffItems;
    }

    /**
     * Compare two PDFs visually using positioned text data
     * @param {Array<PageTextContent>} pdf1Content - First PDF's positioned text content
     * @param {Array<PageTextContent>} pdf2Content - Second PDF's positioned text content
     * @returns {VisualDiffResult} Visual diff result with positioned changes
     */
    function compareVisualPdfs$1(pdf1Content, pdf2Content) {
      const allDiffItems = [];
      const maxPages = Math.max(pdf1Content.length, pdf2Content.length);
      
      let totalAdded = 0;
      let totalRemoved = 0;
      let totalUnchanged = 0;
      
      // Compare each page
      for (let pageNum = 1; pageNum <= maxPages; pageNum++) {
        const page1 = pdf1Content.find(p => p.page === pageNum);
        const page2 = pdf2Content.find(p => p.page === pageNum);
        
        const items1 = page1 ? page1.items : [];
        const items2 = page2 ? page2.items : [];
        
        const matchingResult = findMatchingItems(items1, items2);
        const pageDiffItems = createVisualDiffItems(matchingResult);
        
        // Count changes for this page
        const pageAdded = pageDiffItems.filter(item => item.type === 'added').length;
        const pageRemoved = pageDiffItems.filter(item => item.type === 'removed').length;
        const pageUnchanged = pageDiffItems.filter(item => item.type === 'unchanged').length;
        
        totalAdded += pageAdded;
        totalRemoved += pageRemoved;
        totalUnchanged += pageUnchanged;
        
        allDiffItems.push(...pageDiffItems);
      }
      
      const identical = totalAdded === 0 && totalRemoved === 0;
      
      return {
        items: allDiffItems,
        summary: {
          addedCount: totalAdded,
          removedCount: totalRemoved,
          unchangedCount: totalUnchanged
        },
        identical,
        pageInfo: {
          maxPages
        }
      };
    }

    /**
     * Format visual diff result for overlay rendering
     * @param {VisualDiffResult} visualDiffResult - Visual diff result
     * @param {Object} [options] - Formatting options
     * @param {boolean} [options.showUnchanged=false] - Whether to include unchanged items
     * @param {Array<number>} [options.pages] - Specific pages to include (default: all)
     * @returns {Object} Formatted overlay data by page
     */
    function formatVisualDiffForOverlay(visualDiffResult, options = {}) {
      const { showUnchanged = false, pages } = options;
      const overlayData = {};
      
      // Filter items based on options
      let itemsToRender = visualDiffResult.items;
      
      if (!showUnchanged) {
        itemsToRender = itemsToRender.filter(item => item.type !== 'unchanged');
      }
      
      if (pages && pages.length > 0) {
        itemsToRender = itemsToRender.filter(item => pages.includes(item.page));
      }
      
      // Group by page
      for (const item of itemsToRender) {
        if (!overlayData[item.page]) {
          overlayData[item.page] = [];
        }
        
        overlayData[item.page].push({
          x: item.x,
          y: item.y,
          width: item.width,
          height: item.height,
          text: item.text,
          type: item.type,
          fontName: item.fontName
        });
      }
      
      return overlayData;
    }

    // Browser-compatible version of pdfdiff
    // This is the primary entry point for browser usage


    /**
     * Extract text from PDF data using pdfjs-dist (browser version)
     * @param {ArrayBuffer|Uint8Array|Buffer} pdfData - PDF data
     * @param {import('./utils/pdfjs-extractor.js').PDFJSLib} [pdfjsLib] - Optional pdfjs-dist instance (defaults to global pdfjsLib)
     * @returns {Promise<string>} Extracted text content
     */
    async function extractPdfText(pdfData, pdfjsLib) {
      try {
        const uint8Array = toUint8Array(pdfData);
        return await extractPdfText$1(uint8Array, pdfjsLib);
      } catch (error) {
        throw new Error(`Failed to extract text from PDF: ${error.message}`);
      }
    }

    /**
     * Compare two PDFs (browser version)
     * @param {ArrayBuffer|Uint8Array|Buffer|string} pdf1 - First PDF data or text content
     * @param {ArrayBuffer|Uint8Array|Buffer|string} pdf2 - Second PDF data or text content
     * @param {import('./types.js').BrowserCompareOptions} [options] - Comparison options
     * @returns {Promise<import('./types.js').DiffResult>} Comparison result
     */
    async function comparePdfs(pdf1, pdf2, options = {}) {
      const { ignoreWhitespace = false, ignoreCase = false, mode = 'char', pdfjsLib } = options;
      
      let text1, text2;
      
      // If inputs are strings, use them directly
      if (typeof pdf1 === 'string' && typeof pdf2 === 'string') {
        text1 = pdf1;
        text2 = pdf2;
      } else {
        // Extract text from PDF data
        text1 = await extractPdfText(/** @type {ArrayBuffer|Uint8Array|Buffer} */ (pdf1), pdfjsLib);
        text2 = await extractPdfText(/** @type {ArrayBuffer|Uint8Array|Buffer} */ (pdf2), pdfjsLib);
      }
      
      return compareTexts(text1, text2, { ignoreWhitespace, ignoreCase, mode });
    }

    /**
     * Extract positioned text from PDF data (browser version)
     * @param {ArrayBuffer|Uint8Array|Buffer} pdfData - PDF data
     * @param {import('./utils/pdfjs-extractor.js').PDFJSLib} [pdfjsLib] - Optional pdfjs-dist instance
     * @returns {Promise<Array<import('./utils/pdfjs-extractor.js').PageTextContent>>} Array of page text content with positions
     */
    async function extractPositionedText(pdfData, pdfjsLib) {
      try {
        const uint8Array = toUint8Array(pdfData);
        return await extractPositionedPdfText(uint8Array, pdfjsLib);
      } catch (error) {
        throw new Error(`Failed to extract positioned text from PDF: ${error.message}`);
      }
    }

    /**
     * Compare two PDFs visually with position information (browser version)
     * @param {ArrayBuffer|Uint8Array|Buffer} pdf1 - First PDF data
     * @param {ArrayBuffer|Uint8Array|Buffer} pdf2 - Second PDF data  
     * @param {Object} [options] - Comparison options
     * @param {import('./utils/pdfjs-extractor.js').PDFJSLib} [options.pdfjsLib] - pdfjs-dist instance for text extraction
     * @returns {Promise<import('./utils/visual-diff.js').VisualDiffResult>} Visual diff result with positioned changes
     */
    async function compareVisualPdfs(pdf1, pdf2, options = {}) {
      const { pdfjsLib } = options;
      
      try {
        // Extract positioned text from both PDFs
        const pdf1Content = await extractPositionedText(pdf1, pdfjsLib);
        const pdf2Content = await extractPositionedText(pdf2, pdfjsLib);
        
        // Perform visual comparison
        const visualDiff = compareVisualPdfs$1(pdf1Content, pdf2Content);
        
        return visualDiff;
      } catch (error) {
        throw new Error(`Failed to compare PDFs visually: ${error.message}`);
      }
    }

    /**
     * @typedef {Object} PDFDiffGlobal
     * @property {function(ArrayBuffer|Uint8Array|Buffer, Object=): Promise<string>} extractPdfText - Extract text from PDF data
     * @property {function(ArrayBuffer|Uint8Array|Buffer, Object=): Promise<Array>} extractPositionedText - Extract positioned text from PDF data
     * @property {function(string, string, Object=): import('./types.js').DiffResult} compareTexts - Compare two text strings
     * @property {function(ArrayBuffer|Uint8Array|Buffer|string, ArrayBuffer|Uint8Array|Buffer|string, Object=): Promise<import('./types.js').DiffResult>} comparePdfs - Compare two PDF files
     * @property {function(ArrayBuffer|Uint8Array|Buffer, ArrayBuffer|Uint8Array|Buffer, Object=): Promise<Object>} compareVisualPdfs - Compare two PDFs visually with position data
     * @property {function(import('./types.js').DiffResult, Object=): Promise<string>} formatDiff - Format diff result for display
     * @property {function(Object, Object=): Object} formatVisualDiffForOverlay - Format visual diff for overlay rendering
     */

    // Create a global object for browser usage
    if (typeof window !== 'undefined') {
      /** @type {PDFDiffGlobal} */ 
      (/** @type {any} */ (window).PDFDiff) = {
        extractPdfText,
        extractPositionedText,
        compareTexts,
        comparePdfs,
        compareVisualPdfs,
        formatDiff,
        formatVisualDiffForOverlay
      };
    }

    var browser = { 
      extractPdfText,
      extractPositionedText, 
      compareTexts, 
      comparePdfs,
      compareVisualPdfs, 
      formatDiff,
      formatVisualDiffForOverlay 
    };

    exports.comparePdfs = comparePdfs;
    exports.compareTexts = compareTexts;
    exports.compareVisualPdfs = compareVisualPdfs;
    exports.default = browser;
    exports.extractPdfText = extractPdfText;
    exports.extractPositionedText = extractPositionedText;
    exports.formatDiff = formatDiff;
    exports.formatVisualDiffForOverlay = formatVisualDiffForOverlay;

    Object.defineProperty(exports, '__esModule', { value: true });

    return exports;

})({});
